datafusion/core/src/datasource/listing/url.rs

```diff
@@ -87,6 +89,41 @@ impl ListingTableUrl {
         }
     }
 
+    /// Perform shell-like path expansions
+    /// * Home directory expansion: "~/test.csv" expands to "/Users/user1/test.csv"
+    /// * Environment variable expansion: "$HOME/$DATA/test.csv" expands to
+    /// "/Users/user1/data/test.csv"
+    fn expand_path_prefix(path: &str) -> Result<String, DataFusionError> {
+        let home_dir = home::home_dir()
+            .unwrap()
+            .into_os_string()
+            .into_string()
+            .unwrap();
+
+        let path = path.replace('~', &home_dir);
+
+        let parts = path.split('/').collect::<Vec<_>>();
+        let mut expanded_parts = Vec::new();
+
+        for part in parts {
+            if let Some(envvar_name) = part.strip_prefix('$') {
```

Comment:
this only works for paths like `/$foo/` right, not `/foo$bar/`?

Reply:
Good point, I will add support for patterns like `/foo$bar/` or `/$foo$bar/`